src/Accelerators/NNPA/Runtime/OMRuntimeNNPA.c

```diff
@@ -0,0 +1,74 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//===-------------------------- OMNNPA.c ---------------------------===//
+//
+// Copyright 2022 The IBM Research Authors.
+//
+// =============================================================================
+//
+// Onnx MLIR NNPA Accelerator Runtime
+//
+//===----------------------------------------------------------------------===//
+
+#include <pthread.h>
+
+#include "zdnn.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+// Define variable that tracks whether an accelerator is initialized or not.
+// Initial value is uninitialized.
+// Name must be OMIsInitAccelX where X=NNPA.
+long OMIsInitAccelNNPA = 0;
+
+// Mutex definitions for init and shutdown serialization. A common set is used
+// for all accelerators. Inner mutex is used to implement a fence around the
+// init/shutdown code, to make sure that all side effects from such operations
+// are completed prior to switching the globally readable OMIsInitAccelNNPA.
+pthread_mutex_t OMOuterMutexForInitShutdownNNPA = PTHREAD_MUTEX_INITIALIZER;
+pthread_mutex_t OMInnerMutexForInitShutdownNNPA = PTHREAD_MUTEX_INITIALIZER;
+
+// Define function that performs the serialization of the initialization as well
+// as set the OMIsInitAccelNNPA to true.
+// Name must be OMInitAccelX where X=NNPA.
+
+void OMInitAccelNNPA() {
+  if (!OMIsInitAccelNNPA) {
+    /* Grab outer mutex. */
+    pthread_mutex_lock(&OMOuterMutexForInitShutdownNNPA);
+    /* Test again in the mutex to see if accelerator is not initialized. */
+    if (!OMIsInitAccelNNPA) {
+      /* Still unitinitialized, get inner mutex to fence init code. */
+      pthread_mutex_lock(&OMInnerMutexForInitShutdownNNPA);
+      /* Actual init. */
+      zdnn_init();
+      /* Release inner mutex, and then set accelerator to initialized. */
+      pthread_mutex_unlock(&OMInnerMutexForInitShutdownNNPA);
+      OMIsInitAccelNNPA = 1;
+    } /* Release outer mutex. */
+    pthread_mutex_unlock(&OMOuterMutexForInitShutdownNNPA);
+  }
+}
+
```

Comment:
A couple of problems with this code:

- not sure what's the purpose of using two locks as one lock would be enough to do the job
- you are setting `OMIsInitAccelNNPA` after calling `zdnn_init()`, which means it can be called multiple times, this may or may not work
- in general, it's typically assumed that init/fini calls can be called only once. Therefore, instead of a 1/0 switch, you need a reference count to handle multiple threads. This is a fairly common code pattern, something like the following,
```
init() {
    pthread_mutex_lock();
    if (refcnt == 0) {
        /* do your setup */
    }
    refcnt++;
    pthread_mutex_unlock();
}

fini() {
    pthread_mutex_lock();
    /* It's a bug that fini is called more times than init */
    assert(refcnt > 0);
    if (refcnt == 1) {
        /* do your teardown */
    }
    refcnt--;
    pthread_mutex_unlock();
}
```

Reply:
@gongsu832 the code is correct and uses 2 locks as an abundance of caution. Here is the scenario that the second lock is protecting:
1. I need to test again within the lock if the device has not been init (the refcnt++ in your example, for me setting the `OMIsInitAccelNNPA=1`.
2. Without a fence between the `zdnn_setup()` and the  `OMIsInitAccelNNPA=1`, it would be possible for another thread to see `OMIsInitAccelNNPA=1` before all of the stores associated with the setup are seen by that other thread (release consistency).
3. While on machines like POWER and Z, a memory fence would be enough, there is no posix memory fence that I know of. However, a mutex is guaranteed to have a fence inside the unlock(), thus my use of an inner lock.

Your code snipet is also not valid for the targeted use. I need to be able to read `OMIsInitAccelNNPA` outside of a lock, because grabbing a lock to just see that the device is initialized would be expensive. Instead, I set `OMIsInitAccelNNPA=1` once I can guarantee that every other threads witnessing `OMIsInitAccelNNPA==1` guarantee that they will see the initialized state. Thus, the code inserted inside the `run_main_graph` and equivalent entry points can simply be
```C++
if (!OMIsInitAccelNNPA) OMInitAccelNNPA()
```

There is no need to maintain a ref count, because init can be done only once, and we do not expect a fini after any init. The shutdown is currently not needed, unless we have to do something for migration of threads to another machine.

I know multithreading locks are subtle, I am fairly sure I have this one down. Always welcoming questions as we need to collectively agree that we have the right solution.

Reply:
> Your code snipet is also not valid for the targeted use

How so? It calls init and fini exactly once regardless of the number of threads. Your code, on the other hand, may call init multiple times. Here is how,

- thread 1 and 2 both check `OMIsInitAccelNNPA` to be 0 so both proceed to lock outer mutex, assume thread 1 gets it so thread 2 blocks
- thread 1 goes through the whole process of locking inner mutex, calll zdnn_init, unlock inner mutex, set `OMIsInitAccelNNPA` to 1, unlock outer mutex
- now thread 2 proceeds and does the whole thing again

> because grabbing a lock to just see that the device is initialized would be expensive

This not true. pthread mutex is completely userspace code typically just a few instructions.

> There is no need to maintain a ref count, because init can be done only once, and we do not expect a fini after any init.

Yes there is. The reference count is for making sure whatever shutdown code is called only once. While zdnn doesn't have a shutdown, I'm assuming in general there might be a shutdown for some accelerator. And only the last thread exiting from run_main_graph (or whatever other entry point that has called init) can call shutdown. Your shutdown code, just like your init code, is flawed and does not guarantee the shutdown is called only once.


> I know multithreading locks are subtle, I am fairly sure I have this one down. Always welcoming questions as we need to collectively agree that we have the right solution.

Which is exactly the reason this type of code tries to be as simple as possible, with as few locks as possible.

Reply:
@gongsu832 

There is no need to call a fini for each init. We only need to call init once (replacing the PRELOAD functionality). Calling fini is just pure overhead. And when the count reach zero, there is still nothing to do. The alternative code pattern is perfectly valid when one need to dynamically track the number of active calls. This is just not needed in our use case.

In all but the first (or a few times, if multiple threads are concurrently seen the device as not being initialized), the only test being done is "read, compare to zero, branch around the function call". Regardless of how inexpensive a mutex is, it will be more than that. 

> The reference count is for making sure whatever shutdown code is called only once. 

There is no shutdown code on zdnn. I confirmed with the authors of the zdnn library. The only use case would be in case of live migration to a different node, or a case where they mentioned some custom env variable were to change. They requested that we don't worry about this for now. 

If you prefer, I can remove the shutdown code at this time. I figured that I could leave it because it might be needed in the future. But if it is confusing, I can remove it. Let me know.

> Which is exactly the reason this type of code tries to be as simple as possible, with as few locks as possible.

Low overhead, no mutex that may serialize parallel threads calling distinct `run_main_graph` in parallel.

Reply:
In typical case, threads are all spawned together so you code will be basically all blocking on the outer mutex. And since nobody can proceed until you release the outer mutex, I fail to see the use of the inner mutex. Perhaps you can give an example of when the inner mutex is useful.

I'm not talking about zdnn. On the contrary, I'm assuming an accelerator in general may need an init and an fini, each to be called exactly once. How would you do it?

Reply:
> In typical case, threads are all spawned together so you code will be basically all blocking on the outer mutex.

correct, there is no way to avoid this, and the code section above is no different. Once the flag is set, though, no threads will ever attempt to grab the mutex.

>  I fail to see the use of the inner mutex. 

Correct, the inner mutex will always succeed. The operation I need is the memory fence in the `pthread_mutex_unlock`. Without a memory fence between the stores in the `zdnn_init()` and the `OMIsInitAccelNNPA=1`, another thread may see the store associated with `OMIsInitAccelNNPA=1` before a store inside the init, and thus see the initialization in an inconsistent state. The memory fence basically enforce the following: "if any thread witness a memory state set AFTER the fence, it is guaranteed to see any of the memory state set before the fence". So it does not guarantee any other threads to see `OMIsInitAccelNNPA==1` (thus forcing them possibly in one more pointless mutex), but if they see `OMIsInitAccelNNPA==1` then they are guaranteed to see all of the stores performed in the init.

I could find out what the men fence intrinsic is for Z machine (not sure if zOS and LoZ use the same...) but out of simplicity, I just decided to use a spurious lock to enforce this.

> On the contrary, I'm assuming an accelerator in general may need an init and an fini, each to be called exactly once. How would you do it? Maybe we have different ideas on how the init/fini are called. I thought init was called on entry to run_main_graph (and other entry points), and fini (if necessary) is called on exit from run_main_graph (and other entry points).

The ref count solution indeed satisfy this description. It is not what is required by the zdnn at this stage. It only requires `zdnn_init()` to be called once prior to any other zdnn calls.

Reply:
> Correct, the inner mutex will always succeed. The operation I need is the memory fence in the `pthread_mutex_unlock`. Without a memory fence between the stores in the `zdnn_init()` and the `OMIsInitAccelNNPA=1`, another thread may see the store associated with `OMIsInitAccelNNPA=1` before a store inside the init, and thus see the initialization in an inconsistent state. The memory fence basically enforce the following: "if any thread witness a memory state set AFTER the fence, it is guaranteed to see any of the memory state set before the fence". So it does not guarantee any other threads to see `OMIsInitAccelNNPA==1` (thus forcing them possibly in one more pointless mutex), but if they see `OMIsInitAccelNNPA==1` then they are guaranteed to see all of the stores performed in the init.

There is no memory fence issue in this case. Memory fence is only needed when instructions are scheduled/executed out of order. Since zdnn_init is not an inline function, the last instruction of of zdnn_init will be a "return" instruction. No compiler/hardware will schedule/execute this "return" instruction and OMIsInitAccelNNPA=1 out of order. So no memory fence is ever needed between zdnn_init and OMIsInitAccelNNPA=1.



> The ref count solution indeed satisfy this description. It is not what is required by the zdnn at this stage. It only requires `zdnn_init()` to be called once prior to any other zdnn calls.

So you are going to write this code again for every accelerator we add? I was assuming that you would have a general init/fini like the refcnt code. Then for each accelerator you add, you simply do init(zdnn_init)/fini(NULL), or init(newdnn_init)/fini(newdnn_fini).

Reply:
I have one example where a refcount is useful. We have two threads running in parallel but not exactly start and finish at the same time (in other words, not coming to the mutex at the same time):
- thread 1 calls init, starts first and runs for a while.
- thread 2 jumps in, calls init while thread 1 is still running. (we have no problem here when using refcount or two locks)
- now thread 2 calls fini and finishes first:
    - if we uses the two locks, thread 2 will call fini and successfully shutdown the accelerator. Thread 1 would fail to use the accelerator then.
    - if we use refcount, fini knows that there is another active use by looking at the refcount. fini only decreases the refcount without shutting down the accelerator.